src/main/java/org/prebid/server/bidder/yieldlab/YieldlabBidder.java

```diff
@@ -231,6 +244,60 @@ private static String getConsentParameter(User user) {
         return ObjectUtils.defaultIfNull(consent, "");
     }
 
+    private static Map<String, String> extractDsaRequestParamsFromBidRequest(BidRequest request) {
+        return Optional.ofNullable(request.getRegs())
+            .map(Regs::getExt)
+            .map(ExtRegs::getDsa)
+            .map(YieldlabBidder::extractDsaRequestParamsFromDsaRegsExtension)
+            .orElseGet(Map::of);
+    }
+
+    private static Map<String, String> extractDsaRequestParamsFromDsaRegsExtension(final ExtRegsDsa dsa) {
+        final HashMap<String, String> dsaRequestParams = new HashMap<>();
+
+        if (dsa.getDsaRequired() != null) {
+            dsaRequestParams.put("dsarequired", dsa.getDsaRequired().toString());
+        }
+
+        if (dsa.getPubRender() != null) {
+            dsaRequestParams.put("dsapubrender", dsa.getPubRender().toString());
+        }
+
+        if (dsa.getDataToPub() != null) {
+            dsaRequestParams.put("dsadatatopub", dsa.getDataToPub().toString());
+        }
+
+        final List<ExtRegsDsaTransparency> dsaTransparency = dsa.getTransparency();
+        if (dsaTransparency != null && !dsaTransparency.isEmpty()) {
+            dsaRequestParams.put("dsatransparency", encodeTransparenciesAsString(dsaTransparency));
+        }
+
+        return dsaRequestParams;
+    }
+
+    private static String encodeTransparenciesAsString(List<ExtRegsDsaTransparency> transparencies) {
+        return transparencies.stream()
+            .filter(YieldlabBidder::transparencyIsValid)
+            .map(YieldlabBidder::encodeTransparency)
+            .collect(Collectors.joining(TRANSPARENCY_TEMPLATE_DELIMITER));
+    }
+
+    private static boolean transparencyIsValid(ExtRegsDsaTransparency transparency) {
+        return transparency.getDomain() != null
+                && transparency.getDsaParams() != null
+                && !transparency.getDsaParams().isEmpty();
+    }
+
+    private static String encodeTransparency(ExtRegsDsaTransparency transparency) {
+        return TRANSPARENCY_TEMPLATE.formatted(transparency.getDomain(),
+            encodeTransparencyParams(transparency.getDsaParams()));
```

Comment:
is it possible the `dsaParams` being null? If yes it's worth being handled

FYI the PBS Go seems to handle it with `json:"dsaparams,omitempty"`

Reply:
yeah, my bad, fair point